test/distributed/org/apache/cassandra/distributed/test/ShortReadProtectionTest.java

```diff
@@ -18,46 +18,309 @@
+     * <p>
+     * See CASSANDRA-13880.
+     * <p>
+     * Replaces Python dtest {@code consistency_test.py:TestConsistency.test_13880()}.
+     */
+    @Test
+    public void testSkinnyTableWithoutLiveRows()
+    {
+        tester.createTable("CREATE TABLE %s (id int PRIMARY KEY)")
+              .allNodes("INSERT INTO %s (id) VALUES (0)")
+              .onlyNode1("DELETE FROM %s WHERE id = 0")
+              .assertRows("SELECT DISTINCT id FROM %s WHERE id = 0")
+              .assertRows("SELECT id FROM %s WHERE id = 0 LIMIT 1");
+    }
+
+    /**
+     * Tests SRP for tables with no clustering columns and with alternated live and deleted rows.
+     * <p>
+     * See CASSANDRA-13747.
+     * <p>
+     * Replaces Python dtest {@code consistency_test.py:TestConsistency.test_13747()}.
+     */
+    @Test
+    public void testSkinnyTableWithLiveRows()
+    {
+        tester.createTable("CREATE TABLE %s (id int PRIMARY KEY)")
+              .allNodes(0, 10, i -> String.format("INSERT INTO %%s (id) VALUES (%d)", i)) // order is 5,1,8,0,2,4,7,6,9,3
+              .onlyNode1("DELETE FROM %s WHERE id IN (1, 0, 4, 6, 3)") // delete every other row
+              .assertRows("SELECT DISTINCT token(id), id FROM %s",
+                          row(token(5), 5), row(token(8), 8), row(token(2), 2), row(token(7), 7), row(token(9), 9));
+    }
+
+    /**
+     * Tests SRP for tables with no clustering columns and with complementary deleted rows.
+     * <p>
+     * See CASSANDRA-13595.
+     * <p>
+     * Replaces Python dtest {@code consistency_test.py:TestConsistency.test_13595()}.
+     */
+    @Test
+    public void testSkinnyTableWithComplementaryDeletions()
+    {
+        tester.createTable("CREATE TABLE %s (id int PRIMARY KEY)")
+              .allNodes(0, 10, i -> String.format("INSERT INTO %%s (id) VALUES (%d)", i)) // order is 5,1,8,0,2,4,7,6,9,3
+              .onlyNode1("DELETE FROM %s WHERE id IN (5, 8, 2, 7, 9)") // delete every other row
+              .onlyNode2("DELETE FROM %s WHERE id IN (1, 0, 4, 6)") // delete every other row but the last one
+              .assertRows("SELECT id FROM %s LIMIT 1", row(3))
+              .assertRows("SELECT DISTINCT id FROM %s LIMIT 1", row(3));
+    }
+
+    /**
+     * Tests SRP when more than one row is missing.
+     * <p>
+     * See CASSANDRA-12872.
+     * <p>
+     * Replaces Python dtest {@code consistency_test.py:TestConsistency.test_12872()}.
+     */
+    @Test
+    public void testMultipleMissedRows()
+    {
+        tester.createTable("CREATE TABLE %s (pk int, ck int, PRIMARY KEY (pk, ck))")
+              .allNodes(0, 4, i -> String.format("INSERT INTO %%s (pk, ck) VALUES (0, %d)", i))
+              .onlyNode1("DELETE FROM %s WHERE pk = 0 AND ck IN (1, 2, 3)",
+                         "INSERT INTO %s (pk, ck) VALUES (0, 5)")
+              .onlyNode2("INSERT INTO %s (pk, ck) VALUES (0, 4)")
+              .assertRows("SELECT ck FROM %s WHERE pk = 0 LIMIT 2", row(0), row(4));
```

Comment:
It isn't consistent, but I've observed this failing on the parameterization `[quorum=false flush=true paging=true coordinator=2]`:

```
java.lang.AssertionError: Expected: [[0],[4]]
Actual: [[0],[3]]
	at org.apache.cassandra.distributed.shared.AssertUtils.fail(AssertUtils.java:193)
	at org.apache.cassandra.distributed.shared.AssertUtils.assertTrue(AssertUtils.java:181)
	at org.apache.cassandra.distributed.shared.AssertUtils.assertRows(AssertUtils.java:70)
	at org.apache.cassandra.distributed.test.ShortReadProtectionTest$Tester.assertRows(ShortReadProtectionTest.java:505)
	at org.apache.cassandra.distributed.test.ShortReadProtectionTest$Tester.access$500(ShortReadProtectionTest.java:418)
	at org.apache.cassandra.distributed.test.ShortReadProtectionTest.testMultipleMissedRows(ShortReadProtectionTest.java:193)
```

Reply:
I think this is due to the lack of specific timestamps on the writes done by the test. If we use the automatic ones there is a chance they end up with the same timestamp messing up with reconciliation. Some of the original dtests were specifying that timestamp, while others were missing it. This was probably not a problem is Python but can be problematic with their faster Java version. Fixed [here](https://github.com/apache/cassandra/pull/801/commits/38d41c3e4e543731762ab121aa4bdb4fd4d428f5).

Reply:
Interesting. Well, in any case, I can't get it to fail now :)